### PR TITLE
Add replacers to inject dynamic content

### PR DESCRIPTION
**Pull request by riasvdv**
This PR introduces the following things:

- A new `replacers` function on the CacheProfile interface
- Replacers will be called before caching a page
- After returning a cached response, the replacer will be called again and the old value will be replaced with the new value
- The BaseCacheProfile implements this for the `csrf_token`
- Updates the docs

Breaking change because the interface has changed.

### CODE REVIEW

File: src/ResponseCacheRepository.php [hunk 1]
```diff
@@ -29,6 +29,11 @@ public function put(string $key, $response, $seconds)
         $this->cache->put($key, $this->responseSerializer->serialize($response), $seconds);
     }
 
+    public function putToken(string $key, $token, $seconds)
```
Comment (freekmurze) [inline src/ResponseCacheRepository.php]: Can we add typehints here?

**Comment by riasvdv** [comment]
I've thought of a better way to implement the replacers, will get back to this

**Comment by riasvdv** [comment]
Restructured this to allow people to implement the ReplacerInterface, we'll provide a default CsrfTokenReplacer in this package. This makes this PR a non-breaking change and replacers can be added in a config setting.

File: src/Replacers/CsrfTokenReplacer.php [hunk 2]
```diff
@@ -0,0 +1,16 @@
+<?php
+
+namespace Spatie\ResponseCache\Replacers;
+
+class CsrfTokenReplacer implements ReplacerInterface
```
Comment (freekmurze) [inline src/Replacers/CsrfTokenReplacer.php]: Let's ditch the `interface` suffix.

File: src/Replacers/CsrfTokenReplacer.php [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+{
+    public function getKey(): string
```
Comment (freekmurze) [inline src/Replacers/CsrfTokenReplacer.php]: Rename this to `searchFor`

File: src/Replacers/CsrfTokenReplacer.php [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+    {
+        return 'csrf-token';
+    }
+
+    public function getValue(): string
```
Comment (freekmurze) [inline src/Replacers/CsrfTokenReplacer.php]: Rename this to `replaceBy`

File: src/ResponseCache.php [hunk 5]
```diff
@@ -54,6 +55,17 @@ public function cacheResponse(Request $request, Response $response, $lifetimeInS
             ($lifetimeInSeconds) ? intval($lifetimeInSeconds) : $this->cacheProfile->cacheRequestUntil($request)
         );
 
+        foreach (config('responsecache.replacers', []) as $replacerClass) {
```
Comment (freekmurze) [inline src/ResponseCache.php]: Let's replace this by a collection loop. In the first `map` operation you can can convert the string to an actual class. In the second `each` operation you can just typehint `Replacer`.

File: src/Middlewares/CacheResponse.php [hunk 6]
```diff
@@ -25,7 +30,19 @@ public function handle(Request $request, Closure $next, $lifetimeInSeconds = nul
             if ($this->responseCache->hasBeenCached($request)) {
                 event(new ResponseCacheHit($request));
 
-                return $this->responseCache->getCachedResponseFor($request);
+                $response = $this->responseCache->getCachedResponseFor($request);
+
+                if ($response->getContent()) {
+                    foreach (config('responsecache.replacers', []) as $replacerClass) {
```
Comment (freekmurze) [inline src/Middlewares/CacheResponse.php]: Replace this by a collection chain.

File: src/Middlewares/CacheResponse.php [hunk 7]
```diff
@@ -25,7 +30,19 @@ public function handle(Request $request, Closure $next, $lifetimeInSeconds = nul
...
+                        $replacer = resolve($replacerClass);
+                        if ($replacer instanceof \Spatie\ResponseCache\Replacers\Replacer) {
```
Comment (freekmurze) [inline src/Middlewares/CacheResponse.php]: Always import namespaces in code.

File: README.md [hunk 8]
```diff
@@ -210,9 +218,36 @@ interface CacheProfile
...
+/*
+ * Here you may define replacers that dynamically replace content from the response.
+ * Each replacer must implement the ReplacerInterface.
+ */
+'replacers' => [
+    \Spatie\ResponseCache\Replacers\CsrfTokenReplacer::class,
+],
+```
+
+The callback will be called when the response is cached for the first time,
```
Comment (freekmurze) [inline README.md]: Which callback are you referring to here?

File: README.md [hunk 9]
```diff
@@ -210,9 +218,36 @@ interface CacheProfile
...
+### Injecting dynamic content with Replacers
+If you need inject dynamic content into the cached response, you can register replacers in the config file.
+We add a `CsrfTokenReplacer` by default in the config file.
+
+Replacers must implement the `Replacer` interface.
+
+```php
+/*
+ * Here you may define replacers that dynamically replace content from the response.
+ * Each replacer must implement the ReplacerInterface.
```
Comment (freekmurze) [inline README.md]: Remove the `interface` suffix here.

File: src/ResponseCache.php [hunk 10]
```diff
@@ -3,12 +3,13 @@
...
 
 use Illuminate\Http\Request;
+use Spatie\ResponseCache\Replacers\Replacer;
 use Symfony\Component\HttpFoundation\Response;
 use Spatie\ResponseCache\CacheProfiles\CacheProfile;
 
 class ResponseCache
 {
-    /** @var \Spatie\ResponseCache\ResponseCache */
+    /** @var \Spatie\ResponseCache\ResponseCacheRepository */
```
Comment (freekmurze) [inline src/ResponseCache.php]: 👍

File: README.md [hunk 11]
```diff
@@ -83,6 +83,14 @@ return [
      * You may use a string or an array here.
      */
     'cache_tag' => '',
+    
+    /*
+     * Here you may define replacers that dynamically replace content from the response.
+     * Each replacer must implement the ReplacerInterface.
```
Comment (freekmurze) [inline README.md]: Remove the `interface` suffix

File: src/Exceptions/InvalidReplacer.php [hunk 12]
```diff
@@ -0,0 +1,13 @@
+<?php
+
+namespace Spatie\ResponseCache\Exceptions;
+
+use Exception;
+
+class InvalidReplacer extends Exception
+{
+    public static function callbackString(): self
```
Comment (freekmurze) [inline src/Exceptions/InvalidReplacer.php]: Remove this to `callbackMustReturnString`

File: src/Exceptions/InvalidReplacer.php [hunk 13]
```diff
@@ -0,0 +1,13 @@
+<?php
+
+namespace Spatie\ResponseCache\Exceptions;
+
+use Exception;
+
+class InvalidReplacer extends Exception
```
Comment (freekmurze) [inline src/Exceptions/InvalidReplacer.php]: Maybe I'm looking over it, but is this one actually being thrown somewhere?

File: README.md (hunk 8 above)
Comment (riasvdv) [reply README.md]: Was from the old implementation

File: src/Exceptions/InvalidReplacer.php (hunk 13 above)
Comment (riasvdv) [reply src/Exceptions/InvalidReplacer.php]: From old implementation, deleted

**Comment by riasvdv** [comment]
See #189

**Pull request closed without merging**